docs/src/design/multiset.md

```diff
@@ -1,56 +1,158 @@
+1. To prove the consistency of intermediate values that must persist between different cycles of the trace, without storing the full data in the execution trace (which would require adding more columns to the trace).
+2. To prove correct interaction between two independent sections of the execution trace, e.g., between the main trace where the result of some operation is required, but would be expensive to compute, and a specialized component which can perform that operation cheaply.
 
-To enable this in the bitwise case, we have a special “Bitwise co-processor” which is our equivalent of a specialized circuit that will return the result of bitwise operations in a way that:
+The first is achieved using [virtual tables](#virtual-tables) of data, where we add a row at some cycle in the trace and remove it at a later cycle when it is needed again. Instead of maintaining the entire table in the execution trace, multiset checks allow us to prove data consistency of this table using one running product column.
 
-1. allows us to enforce constraints and prove the correctness of the result for the given inputs.
-2. allows us to do this more efficiently then if we actually had to execute this without a specialized design.
+The second is done by reducing each operation to a lookup value and then using a [communication bus](#communication-buses), implemented as a running product column, to provably connect the two sections of the trace.
 
-The co-processor takes in the two inputs and the bitwise operation and then provably computes the result so we could just look it up, which is exactly what we need(More information on [bitwise design doc](./chiplets/bitwise.md)).
+### Running product columns
+Although the multiset equality check can be thought of as comparing multiset equality between two vectors $a$ and $b$, in Miden VM it is implemented as a single running product column in the following way:
 
-But now we’re left with a problem - we can prove that the bitwise computation is done correctly, but it’s not connected to our main processor yet, so we still need a way to actually “look up” the value and prove that the value we’re using after our lookup is in fact the correct result from that co-processor for the inputs we specified.
-
-**NOTE** - The stack and the lookup value might not necessarily be sharing the same trace. In fact, the likelihood of both sharing the same/adjacent row is very low. Given STARK constraints are local(constraints can be applied only on two consecutive traces), we can't enforce any transition constraint here.
-
-**Permutation checks are what allow us to link these different processors and prove that this lookup is being executed correctly.**
+- The running product column is initialized to $1$ at the beginning of the trace.
```

Comment:
This is not always true (i.e., sometimes we can initialize a column with some other value) - but probably doesn't make sense to overcomplicate it here with this nuance.

Reply:
Yeah, I just wanted to keep it simple when describing how we do things, since they can look at the multiset checks write-up for a more detailed explanation. I'll see if there's a way to adjust this to be a bit more exact without adding confusion though.

Reply:
Sounds good! But again, if there isn't a good way - it is fine to keep as is.